pkg/handler/collector/oci/oci.go

```diff
@@ -183,89 +208,192 @@ func (o *ociCollector) getTagsAndFetch(ctx context.Context, repo string, tags []
 		if !contains(o.checkedDigest[repo], digestTag) {
 			imageTag := fmt.Sprintf("%v:%v", repo, digestTag)
-			r, err := ref.New(imageTag)
+			err = fetchOCIArtifactBlobs(ctx, rc, imageTag, "unknown", docChannel)
 			if err != nil {
-				return err
+				return fmt.Errorf("failed retrieving artifact blobs from registry fallback artifacts: %w", err)
 			}
+			o.checkedDigest[repo] = append(o.checkedDigest[repo], digestTag)
+		}
+	}
 
-			// if `.att` or `.sbom`` do not exist for specified digest
-			// log error and continue
-			m, err = rc.ManifestGet(ctx, r)
-			if err != nil {
-				// this is a normal behavior, not an error when the digest does not have an attestation
-				// explicitly logging it as info to avoid call-stack when logging
-				logger.Infof("unable to get manifest for %v: %v", imageTag, err)
-				continue
-			}
+	referrerList, err := rc.ReferrerList(ctx, image)
+	if err != nil {
+		return fmt.Errorf("failed retrieving referrer list: %w", err)
+	}
 
-			// go through layers in reverse
-			mi, ok := m.(manifest.Imager)
-			if !ok {
-				return fmt.Errorf("reference is not a known image media type")
-			}
-			layers, err := mi.GetLayers()
-			if err != nil {
-				return err
-			}
-			for i := len(layers) - 1; i >= 0; i-- {
-				if ctx.Err() != nil {
-					return ctx.Err()
-				}
-				blob, err := rc.BlobGet(ctx, r, layers[i])
-				if err != nil {
-					return fmt.Errorf("failed pulling layer %d: %w", i, err)
-				}
-				btr1, err := blob.RawBody()
-				if err != nil {
-					return err
-				}
+	logger.Infof("Found %d referrers for %s", len(referrerList.Descriptors), image.Digest)
+
+	// Use goroutines to fetch referrers concurrently
+	// Create a channel to collect errors from goroutines
+	errorChan := make(chan error, len(referrerList.Descriptors))
+	var wg sync.WaitGroup
 
-				doc := &processor.Document{
-					Blob:   btr1,
-					Type:   processor.DocumentUnknown,
-					Format: processor.FormatUnknown,
-					SourceInformation: processor.SourceInformation{
-						Collector: string(OCICollector),
-						Source:    imageTag,
-					},
+	for _, referrerDesc := range referrerList.Descriptors {
+		// Increment the WaitGroup counter
+		wg.Add(1)
+		go func(referrerDesc types.Descriptor) {
+			defer wg.Done() // Decrement the WaitGroup counter when done
+			if _, ok := wellKnownOCIArtifactTypes[referrerDesc.ArtifactType]; ok {
+				referrerDescDigest := referrerDesc.Digest.String()
+
+				if !contains(o.checkedDigest[repo], referrerDescDigest) {
+					logger.Infof("Fetching referrer %s with artifact type %s", referrerDescDigest, referrerDesc.ArtifactType)
+					referrerDigest := fmt.Sprintf("%v@%v", repo, referrerDescDigest)
+					err = fetchOCIArtifactBlobs(ctx, rc, referrerDigest, referrerDesc.ArtifactType, docChannel)
+					if err != nil {
+						errorChan <- fmt.Errorf("failed retrieving artifact blobs from registry: %w", err)
+					}
+					o.checkedDigest[repo] = append(o.checkedDigest[repo], referrerDescDigest)
```

Comment:
This is appending to the list across multiple go routines, which is a race condition bug.

Reply:
Thanks, I realized this when my tests failed against the [data race detector](https://go.dev/doc/articles/race_detector). I refactored to use `sync.Map` for this.